File: app/src/main/java/com/grace/app/injection/module/AppModule.java
package com.grace.app.injection.module;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.support.annotation.NonNull;

import com.bumptech.glide.Glide;
import com.bumptech.glide.RequestManager;
import com.grace.app.GraceApplication;
import com.grace.app.model.GraceApiCall;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

@Module
public final class AppModule {
    @NonNull
    private final GraceApplication mGraceApplication;

    public AppModule(@NonNull GraceApplication graceApplication) {
        mGraceApplication = graceApplication;
    }

    @Provides
    public Context provideAppContext() {
        return mGraceApplication;
    }

    @Provides
    public GraceApplication provideApp() {
        return mGraceApplication;
    }

    @Provides
    @Singleton
    SharedPreferences providesSharedPreferences(GraceApplication application) {
        return PreferenceManager.getDefaultSharedPreferences(application);
    }

    @Provides
    @Singleton
    RequestManager provideGlide(GraceApplication application) {
        return Glide.with(application);
    }


    @Provides
    @Singleton
    GraceApiCall provideGraceApiCall() {
        return new GraceApiCall();
    }

}

File: app/src/main/java/com/grace/app/di/AppModule.kt
package com.grace.app.di

import android.content.Context
import android.content.SharedPreferences
import androidx.preference.PreferenceManager
import com.grace.app.custom.FoodDetector
import com.grace.app.data.MealRepository
import com.grace.app.data.MealRepositoryImpl
import com.grace.app.preferences.AppPreferences
import com.grace.app.preferences.SharedPreferencesAppPreferences
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.android.qualifiers.ApplicationContext
import dagger.hilt.components.SingletonComponent
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import javax.inject.Qualifier
import javax.inject.Singleton

@Retention(AnnotationRetention.BINARY)
@Qualifier
annotation class IoDispatcher

@Module
@InstallIn(SingletonComponent::class)
object AppModule {
    @Provides
    @Singleton
    fun provideSharedPreferences(@ApplicationContext context: Context): SharedPreferences {
        return PreferenceManager.getDefaultSharedPreferences(context)
    }

    @Provides
    @Singleton
    fun provideFoodDetector(): FoodDetector = FoodDetector()

    @Provides
    @Singleton
    fun provideMealRepository(impl: MealRepositoryImpl): MealRepository = impl

    @Provides
    @Singleton
    fun provideAppPreferences(preferences: SharedPreferences): AppPreferences {
        return SharedPreferencesAppPreferences(preferences)
    }

    @Provides
    @IoDispatcher
    fun provideIoDispatcher(): CoroutineDispatcher = Dispatchers.IO
}
